src/components/TodoItem.tsx

```diff
@@ -0,0 +1,128 @@
+/* eslint-disable jsx-a11y/label-has-associated-control */
+import React, { useEffect, useRef, useState } from 'react';
+import { Todo } from '../types/Todo';
+import cn from 'classnames';
+
+type Props = {
+  todo: Todo;
+  handleDeletingTodo?: () => void;
+  isLoading: boolean;
+  handleUpdatingTodo?: (todo: Todo) => void;
+  handleTogglingTodo?: () => void;
+};
+
+export const TodoItem: React.FC<Props> = ({
+  todo,
+  handleDeletingTodo = () => {},
+  isLoading,
+  handleUpdatingTodo = () => {},
+  handleTogglingTodo = () => {},
+}) => {
+  const [isEditing, setIsEditing] = useState(false);
+  const [newTitle, setNewTitle] = useState(todo.title);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (isEditing && inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [isEditing]);
+
+  const handleSubmittingUpdate = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    const preparedTitle = newTitle.trim();
+
+    setNewTitle(preparedTitle);
+
+    if (preparedTitle === todo.title) {
+      setIsEditing(false);
+
+      return;
+    }
+
+    if (!preparedTitle) {
+      handleDeletingTodo();
+
+      return;
+    }
+
+    handleUpdatingTodo({ ...todo, title: preparedTitle });
+    setIsEditing(false);
+  };
+
+  return (
+    <div
+      data-cy="Todo"
+      className={cn('todo', {
+        completed: todo.completed,
+      })}
+      key={todo.id}
+    >
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={todo.completed}
+          onChange={handleTogglingTodo}
+        />
+      </label>
+
+      {isEditing ? (
+        <form
+          onSubmit={handleSubmittingUpdate}
+          onKeyUp={event => {
+            if (event.key === 'Escape') {
+              setIsEditing(false);
+            }
+          }}
```

Comment:
Pls move the function inside `onKeyUp` handler to a separate const and use as a callback here
**PLS DO THE SAME FOR ALL SUCH CASES IN THE PROJECT**